### PR TITLE
feat: type support for timestamp in json to moonlink row converter

### PR DESCRIPTION
**Pull request by peterxcli**
## Summary

Type support for timestamp in json to moonlink row converter

## Related Issues

Closes https://github.com/Mooncake-Labs/moonlink/issues/1284

## Changes

- Add a datatime_utils file with different time string parsing and basic timestamp testing
- make json converter to support Date32, Time64 and Timetamp and add some edge case testing

## Checklist

- [x] Code builds correctly
- [x] Tests have been added or updated
- [x] Documentation updated if necessary
- [x] I have reviewed my own changes

### CODE REVIEW

File: src/moonlink_connectors/src/rest_ingest/json_converter.rs [hunk 1]
```diff
@@ -82,7 +83,30 @@ impl JsonToMoonlinkRowConverter {
                     Err(JsonToMoonlinkRowError::TypeMismatch(field.name().clone()))
                 }
             }
-            // TODO: Add more type conversions (Bool, Float, Date, etc.)
```
Comment (peterxcli) [inline src/moonlink_connectors/src/rest_ingest/json_converter.rs:85]: Removed cuz mentioned types are all supported.

File: src/moonlink_connectors/src/rest_ingest/json_converter.rs [hunk 2]
```diff
@@ -82,7 +83,30 @@ impl JsonToMoonlinkRowConverter {
...
+            }
+            DataType::Time64(TimeUnit::Microsecond) => {
+                if let Some(s) = value.as_str() {
+                    parse_time(s)
+                        .map_err(|_| JsonToMoonlinkRowError::InvalidValue(field.name().clone()))
+                } else {
+                    Err(JsonToMoonlinkRowError::TypeMismatch(field.name().clone()))
+                }
+            }
+            DataType::Timestamp(TimeUnit::Microsecond, _tz) => {
```
Comment (dentiny) [inline src/moonlink_connectors/src/rest_ingest/json_converter.rs]: I would like to keep the same behavior as pg -> moonlink type:
- pg replicate uses UTC timezone: https://github.com/Mooncake-Labs/moonlink/blob/70f102a173cbf8cc536d7d8acfac3e0126ad614a/src/moonlink_connectors/src/pg_replicate/conversions.rs#L29
- get integer value we get is also after canonicalizing to UTC timezone: https://github.com/Mooncake-Labs/moonlink/blob/70f102a173cbf8cc536d7d8acfac3e0126ad614a/src/moonlink_connectors/src/pg_replicate/util.rs#L272-L285

File: src/moonlink_connectors/src/rest_ingest/datetime_utils.rs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+/// Parse a time string in HH:MM:SS[.fraction] format to Time64 (microseconds since midnight)
+pub fn parse_time(time_str: &str) -> Result<RowValue, String> {
+    // Try parsing with fractional seconds first, then without
+    let time = NaiveTime::parse_from_str(time_str, "%H:%M:%S%.f")
+        .or_else(|_| NaiveTime::parse_from_str(time_str, "%H:%M:%S"))
+        .map_err(|e| format!("Invalid time format: {e}"))?;
+
+    // Convert to microseconds since midnight
+    let duration = time.signed_duration_since(NaiveTime::from_hms_opt(0, 0, 0).unwrap());
```
Comment (dentiny) [inline src/moonlink_connectors/src/rest_ingest/datetime_utils.rs]: nit: make midnight a constant, just like what you did for UNIX timestamp

File: src/moonlink_connectors/src/rest_ingest/datetime_utils.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    let microseconds = duration
+        .num_microseconds()
+        .ok_or_else(|| "Time value too large".to_string())?;
```
Comment (dentiny) [inline src/moonlink_connectors/src/rest_ingest/datetime_utils.rs]: include error also in the string?

File: src/moonlink_connectors/src/rest_ingest/datetime_utils.rs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_parse_date() {
+        // Valid date
+        let result = parse_date("2024-03-15").unwrap();
+        let expected_days = NaiveDate::from_ymd_opt(2024, 3, 15)
```
Comment (dentiny) [inline src/moonlink_connectors/src/rest_ingest/datetime_utils.rs]: nit: the test code is almost a copy & paste to production code, not sure how valuable it is.
What about using a constant directly? Just like what you did at `test_parse_time`

File: src/moonlink_connectors/src/rest_ingest/json_converter.rs (hunk 2 above)
Comment (dentiny) [reply src/moonlink_connectors/src/rest_ingest/json_converter.rs]: Update:
- I see you handle timestamp conversion to UTC in string parsing part, but it's not guaranteed that timezone information has to be there
- Example: https://github.com/apache/arrow-rs/blob/04f217b6708eed2804c3b0a669a65ea111c2c5f1/arrow-schema/src/datatype_parse.rs#L704-L705

File: src/moonlink_connectors/src/rest_ingest/json_converter.rs [hunk 6]
```diff
@@ -106,6 +131,23 @@ mod tests {
         ]))
     }
 
+    fn make_datetime_schema() -> Arc<Schema> {
+        Arc::new(Schema::new(vec![
+            Field::new("date", DataType::Date32, false),
```
Comment (dentiny) [inline src/moonlink_connectors/src/rest_ingest/json_converter.rs]: nit: add a comment on constant variables
```suggestion
            Field::new("date", DataType::Date32, /*nullable=*/false),
```
Same for all below

File: src/moonlink_connectors/src/rest_ingest/datetime_utils.rs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        // Time with fractional seconds
+        let result = parse_time("09:15:30.123456").unwrap();
+        assert_eq!(result, RowValue::Int64(33330123456));
+
+        // Invalid time
+        assert!(parse_time("25:00:00").is_err());
+    }
+
+    #[test]
+    fn test_parse_timestamp() {
```
Comment (dentiny) [inline src/moonlink_connectors/src/rest_ingest/datetime_utils.rs:114]: Please add a unit test with no timezone specified

File: src/moonlink_connectors/src/rest_ingest/json_converter.rs [hunk 8]
```diff
@@ -212,4 +254,185 @@ mod tests {
...
+        });
+        let err = converter.convert(&input).unwrap_err();
+        match err {
+            JsonToMoonlinkRowError::InvalidValue(f) => assert_eq!(f, "timestamp"),
+            _ => panic!("unexpected error: {err:?}"),
+        }
+    }
+
+    #[test]
+    fn test_edge_cases() {
```
Comment (dentiny) [inline src/moonlink_connectors/src/rest_ingest/json_converter.rs]: nit:
```suggestion
    fn test_datetime_edge_cases() {
```

File: src/moonlink_connectors/src/rest_ingest/json_converter.rs [hunk 9]
```diff
@@ -212,4 +254,185 @@ mod tests {
...
+        let row = converter.convert(&input).unwrap();
+
+        let expected_days = NaiveDate::from_ymd_opt(2024, 2, 29)
+            .unwrap()
+            .signed_duration_since(NaiveDate::from_ymd_opt(1970, 1, 1).unwrap())
+            .num_days() as i32;
+        assert_eq!(row.values[0], RowValue::Int32(expected_days));
+
+        // 23:59:59.999999 = 86399999999 microseconds
+        assert_eq!(row.values[1], RowValue::Int64(86399999999));
```
Comment (dentiny) [inline src/moonlink_connectors/src/rest_ingest/json_converter.rs:525]: check for all row values?

File: src/moonlink_connectors/src/rest_ingest/json_converter.rs [hunk 10]
```diff
@@ -212,4 +254,185 @@ mod tests {
...
+        let input = json!({
+            "date": "1970-01-01",
+            "time": "00:00:00",
+            "timestamp": "1970-01-01T00:00:00Z",
+            "timestamp_utc": "1970-01-01T00:00:00Z",
+        });
+        let row = converter.convert(&input).unwrap();
+        assert_eq!(row.values[0], RowValue::Int32(0)); // 0 days since epoch
+        assert_eq!(row.values[1], RowValue::Int64(0)); // 0 microseconds since midnight
+        assert_eq!(row.values[2], RowValue::Int64(0)); // 0 microseconds since epoch
```
Comment (dentiny) [inline src/moonlink_connectors/src/rest_ingest/json_converter.rs:506]: check all row values?

File: src/moonlink_connectors/src/rest_ingest/json_converter.rs (hunk 2 above)
Comment (peterxcli) [reply src/moonlink_connectors/src/rest_ingest/json_converter.rs]: yeah, I added https://github.com/Mooncake-Labs/moonlink/pull/1336/commits/9390b82b35f2d4833e132a54782f9d9e3a7ad009 for this

**Comment by peterxcli** [comment]
@dentiny Thanks so much for the prompt and thorough review. Addressed all of you comments. Please take another look.

File: src/moonlink_connectors/src/rest_ingest/json_converter.rs (hunk 2 above)
Comment (dentiny) [inline src/moonlink_connectors/src/rest_ingest/json_converter.rs]: Let me make myself clear:
- It's not guaranteed that timezone information is stored within `TimeUnit::Microsecond`, there's another timezone field
- My expected implementation is convert timestamp to UTC timezone, and call parsing function

Does it make sense to you?

Comment (peterxcli) [reply src/moonlink_connectors/src/rest_ingest/json_converter.rs]: Cause they are all string, so I have to write a parsing function specially for this case.

see https://github.com/Mooncake-Labs/moonlink/pull/1336/commits/5dbe0a0eefc5daf8803f6de68de1ff657dd7f4d8

File: src/moonlink_connectors/src/rest_ingest/json_converter.rs [hunk 11]
```diff
@@ -99,9 +99,9 @@ impl JsonToMoonlinkRowConverter {
                     Err(JsonToMoonlinkRowError::TypeMismatch(field.name().clone()))
                 }
             }
-            DataType::Timestamp(TimeUnit::Microsecond, _tz) => {
+            DataType::Timestamp(TimeUnit::Microsecond, tz) => {
                 if let Some(s) = value.as_str() {
-                    parse_timestamp(s)
+                    parse_timestamp_with_timezone(s, tz.as_deref())
```
Comment (dentiny) [inline src/moonlink_connectors/src/rest_ingest/json_converter.rs:104]: (no action needed) our current error propagation actually loses conversion failure information

**Comment by peterxcli** [comment]
Thanks for reviewing!

File: src/moonlink_connectors/src/rest_ingest/datetime_utils.rs [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    let naive_dt = NaiveDateTime::parse_from_str(timestamp_str, "%Y-%m-%dT%H:%M:%S")
+        .or_else(|_| NaiveDateTime::parse_from_str(timestamp_str, "%Y-%m-%dT%H:%M:%S%.f"))
+        .map_err(|e| format!("Invalid timestamp format: {e}"))?;
+
+    // If schema specifies a timezone, interpret the naive datetime in that timezone
+    if let Some(tz_str) = schema_timezone {
+        if let Ok(tz) = tz_str.parse::<chrono_tz::Tz>() {
+            let local_dt = tz
+                .from_local_datetime(&naive_dt)
+                .single()
```
Comment (dentiny) [inline src/moonlink_connectors/src/rest_ingest/datetime_utils.rs:71]: I think this is buggy, this test case fails:
```rs
parse_timestamp_with_timezone(
        "2024-11-03T01:30:00",
        Some("America/New_York"),
    )
```
I will make a small PR to fix it.

Comment (dentiny) [reply src/moonlink_connectors/src/rest_ingest/datetime_utils.rs:71]: The error message is not accurate as well, since it's also possible to have cases where "the local time does not exist".
https://github.com/chronotope/chrono/blob/b4b9aa580cc61354d5a1a44d7eb881f29ef39fc3/src/offset/mod.rs#L80-L95

I will update it also in the followup PR.

**Comment by dentiny** [comment]
FYI, followup PR: https://github.com/Mooncake-Labs/moonlink/pull/1383
